### PR TITLE
feat: add analysis functions

### PR DESCRIPTION
**Pull request by strixy16**
Includes correlation calculations and plotting those correlations as a heatmap and histogram

### CODE REVIEW

File: src/readii/analyze/correlation.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    vertical_features : pd.DataFrame
+        Dataframe containing features to calculate correlations with. Index must be the same as the index of the horizontal_features dataframe.
+    horizontal_features : pd.DataFrame
+        Dataframe containing features to calculate correlations with. Index must be the same as the index of the vertical_features dataframe.
+    method : str
+        Method to use for calculating correlations. Default is "pearson".
+    vertical_feature_name : str
+        Name of the vertical features to use as suffix in correlation dataframe. Default is blank "".
+    horizontal_feature_name : str
+        Name of the horizontal features to use as suffix in correlation dataframe. Default is blank "".
```
Comment (jjjermiah) [inline src/readii/analyze/correlation.py:22]:
```suggestion
    vertical_feature_name : str
        Name of the vertical features to use as suffix in correlation dataframe. Default is "_vertical".
    horizontal_feature_name : str
        Name of the horizontal features to use as suffix in correlation dataframe. Default is "_horizontal".
```

File: src/readii/analyze/correlation.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
+import pandas as pd
+from readii.utils import logger
+
+def getFeatureCorrelations(vertical_features:pd.DataFrame,
+                           horizontal_features:pd.DataFrame,
+                           method:str = "pearson",
+                           vertical_feature_name:str = '_vertical',
+                           horizontal_feature_name:str = '_horizontal') -> pd.DataFrame:
+    """ Function to calculate correlation between two sets of features.
```
Comment (jjjermiah) [inline src/readii/analyze/correlation.py:9]:
```suggestion
    """Calculate correlation between two sets of features.
```

File: src/readii/analyze/correlation.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        logger.exception(msg)
+        raise e
+
+    return correlation_matrix
+
+
+
+def getVerticalSelfCorrelations(correlation_matrix:pd.DataFrame,
+                                num_vertical_features:int) -> pd.DataFrame:
+    """ Function to get the vertical (y-axis) self correlations from a correlation matrix. Gets the top left quadrant of the correlation matrix.
```
Comment (jjjermiah) [inline src/readii/analyze/correlation.py:76]:
```suggestion
    """Get the vertical (y-axis) self correlations from a correlation matrix. 
    
    Gets the top left quadrant of the correlation matrix.
```

File: src/readii/analyze/correlation.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        raise ValueError()
+
+    # Get the correlation matrix for vertical vs vertical - this is the top left corner of the matrix
+    return correlation_matrix.iloc[0:num_vertical_features, 0:num_vertical_features]
+
+
+
+def getHorizontalSelfCorrelations(correlation_matrix:pd.DataFrame,
+                                  num_horizontal_features:int) -> pd.DataFrame:
+    """ Function to get the horizontal (x-axis) self correlations from a correlation matrix. Gets the bottom right quadrant of the correlation matrix.
```
Comment (jjjermiah) [inline src/readii/analyze/correlation.py:107]:
```suggestion
    """Get the horizontal (x-axis) self correlations from a correlation matrix. 
    
    Gets the bottom right quadrant of the correlation matrix.
```

File: src/readii/io/writers/plot_writer.py
Comment (jjjermiah) [inline src/readii/io/writers/plot_writer.py:1]: this looks dope and very helpful 

if you know what the viable extensions matplotlib allows, you could validate the `self.filename_format` in the `__post__init__` to raise an error like https://github.com/bhklab/readii/blob/36bf8e307a1fc04478170f243350a893f32f5076/src/readii/io/writers/nifti_writer.py#L48-L65

File: src/readii/io/writers/plot_writer.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        logger.debug("Saving.", kwargs=kwargs)
+
+        # Generate the output path
+        out_path = self.resolve_path(**kwargs)
+
+        # Check if the output path already exists
+        if out_path.exists():
+            if not self.overwrite:
+                msg = f"File {out_path} already exists. \nSet {self.__class__.__name__}.overwrite to True to overwrite."
+                logger.exception(msg)
```
Comment (jjjermiah) [inline src/readii/io/writers/plot_writer.py:43]: the logger.exception is most useful for the main process that calls `writer.save`, as it might print out duplicated stack-traces

File: src/readii/analyze/plot_correlation.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    bin_edges : np.ndarray
+        Numpy array containing the bin edges for the histogram.
+    """
+    
+    # Convert to numpy to use histogram function
+    feature_correlation_arr = correlation_matrix.to_numpy()
+
+    # Check if matrix is symmetric
+    if issymmetric(feature_correlation_arr):
+        print("Correlation matrix is symmetric.")
```
Comment (jjjermiah) [inline src/readii/analyze/plot_correlation.py:144]:
```suggestion
        logger.debug("Correlation matrix is symmetric.")
```

File: src/readii/analyze/correlation.py
Comment (jjjermiah) [inline src/readii/analyze/correlation.py:1]: smaller nitpicks 

- function/method docstring one-liner should be [imperative form](https://peps.python.org/pep-0257/#one-line-docstrings), without the `Function to..` prefix
tldr: phrase the one-liner as a command

File: src/readii/analyze/plot_correlation.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    """Function to plot a correlation heatmap.
+
+    Parameters
+    ----------
+    correlation_matrix_df : pd.DataFrame
+        Dataframe containing the correlation matrix to plot.
+    diagonal : bool, optional
+        Whether to only plot half of the matrix. The default is False.
+    triangle : str, optional
+        Which triangle half of the matrixto plot. The default is "lower".
```
Comment (jjjermiah) [inline src/readii/analyze/plot_correlation.py:31]:
```suggestion
    triangle : str, optional
        Which triangle half of the matrix to plot. Either "lower" or "upper".
        The default is "lower".
```

**Review by jjjermiah — CHANGES_REQUESTED** [review]
